### PR TITLE
Improve stair generation, fixes #94

### LINKED ISSUE #94: Lich Tower stairs only use Slabs half:bottom
The stairs in the Lich tower only generate with the "half:bottom" variant of Slabs and uses full blocks for every second step.
I believe this choice was made because Minecraft, at the time, did not yet support half:top Slabs.
Now that this is a possibility, it should be considered changing the full blocks for the half:top variant Slabs.

### PR DESCRIPTION
**Pull request by thehobojoe**
Use top slabs instead of double slab/full block in most cases. Start general refactor of generation, lots of code can be reused in util functions. Obviously, tons more to do here.

### CODE REVIEW

File: src/main/java/twilightforest/structures/StructureTFHelper.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    public <T extends Block> IBlockState getSlab(T type) {
+        return getSlabType(type, BlockSlab.EnumBlockHalf.BOTTOM);
+    }
+
+    public <T extends Block> IBlockState getSlabTop(T type) {
+        return getSlabType(type, BlockSlab.EnumBlockHalf.TOP);
+    }
+
```
Comment (williewillus) [inline src/main/java/twilightforest/structures/StructureTFHelper.java]: The generics here are unnecessary I'm pretty sure. Just use `Block` as the parameter

File: src/main/java/twilightforest/structures/StructureTFComponent.java [hunk 2]
```diff
@@ -39,6 +39,8 @@
 	public StructureTFDecorator deco = null;
 	public int spawnListIndex = 0;
 
+	public StructureTFHelper helper = new StructureTFHelper();
+
```
Comment (williewillus) [inline src/main/java/twilightforest/structures/StructureTFComponent.java]: If this doesn't hold any data can we make its methods static?

**Review by williewillus — CHANGES_REQUESTED** [review]
Thanks for getting rid of all those extra prints and @params :P

File: src/main/java/twilightforest/structures/StructureTFHelper.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    public final IBlockState stoneSlabTop = getSlabTop(Blocks.STONE_SLAB);
+    public final IBlockState stoneSlabDouble = getBlock(Blocks.DOUBLE_STONE_SLAB);
+
+    public IBlockState birchSlab = getSlab(Blocks.WOODEN_SLAB).withProperty(BlockPlanks.VARIANT, BlockPlanks.EnumType.BIRCH);
+    public IBlockState birchSlabTop = getSlabTop(Blocks.WOODEN_SLAB).withProperty(BlockPlanks.VARIANT, BlockPlanks.EnumType.BIRCH);
+    public IBlockState birchPlanks = Blocks.PLANKS.getDefaultState().withProperty(BlockPlanks.VARIANT, BlockPlanks.EnumType.BIRCH);
+
+    public <T extends Block> IBlockState getBlock(T type) {
+        return type.getDefaultState();
+    }
```
Comment (williewillus) [inline src/main/java/twilightforest/structures/StructureTFHelper.java]: Not sure if this is necessary vs just using getDefaultState direvtly.

File: src/main/java/twilightforest/structures/StructureTFHelper.java (hunk 1 above)
Comment (jsulli) [reply src/main/java/twilightforest/structures/StructureTFHelper.java]: Definitely not strictly necessary, it just results in slightly more readable code imo. Can be removed if necessary.

Comment (thehobojoe) [reply src/main/java/twilightforest/structures/StructureTFHelper.java]: oops wrong account

File: src/main/java/twilightforest/structures/StructureTFComponent.java (hunk 2 above)
Comment (thehobojoe) [reply src/main/java/twilightforest/structures/StructureTFComponent.java]: Yep, definitely.

File: src/main/java/twilightforest/structures/StructureTFHelper.java (hunk 3 above)
Comment (thehobojoe) [reply src/main/java/twilightforest/structures/StructureTFHelper.java]: Same answer as above but.... yeah this is definitely the least necessary. I'll probably remove it.

File: src/main/java/twilightforest/structures/StructureTFHelper.java (hunk 1 above)
Comment (thehobojoe) [reply src/main/java/twilightforest/structures/StructureTFHelper.java]: Ah I read what you wrote again - looks like you're right. Not sure why I thought I needed a generic there but apparently I don't. Doing some cleanup now.

**Comment by Benimatic** [comment]
Just bumping this. Is this clear to be merged, williewillus?

**Comment by williewillus** [comment]
lgtm

**Pull request merged**